### PR TITLE
Added lastChild and leftSibling pointer for JsonNodeTrx

### PR DESCRIPTION
**Pull request by Raj-Datta-Manohar**
Resolves issues #294 and #255

### CODE REVIEW

**Comment by Raj-Datta-Manohar** [comment]
@JohannesLichtenberger , Take a look at this PR:-)
I will raise a seperate PR for tests, a little busy this week.

File: bundles/sirix-core/src/main/java/org/sirix/node/delegates/StructNodeDelegate.java [hunk 1]
```diff
@@ -154,15 +197,22 @@ public void incrementChildCount() {
 
   @Override
   public int hashCode() {
-    return Objects.hashCode(childCount, nodeDelegate, firstChild, leftSibling, rightSibling, descendantCount);
+    return ((lastChild==-2L) ? Objects.hashCode(
```
Comment (JohannesLichtenberger) [inline bundles/sirix-core/src/main/java/org/sirix/node/delegates/StructNodeDelegate.java:200]: What's this -2L key?

File: bundles/sirix-core/src/main/java/org/sirix/node/delegates/StructNodeDelegate.java [hunk 2]
```diff
@@ -154,15 +197,22 @@ public void incrementChildCount() {
...
+      childCount, nodeDelegate, firstChild, lastChild, leftSibling, rightSibling, descendantCount));
   }
 
   @Override
   public BigInteger computeHash() {
     final Funnel<StructNode> nodeFunnel = (StructNode node, PrimitiveSink into) ->
     {
-      into.putLong(node.getChildCount()).putLong(node.getDescendantCount()).putLong(node.getLeftSiblingKey()).putLong(
-          node.getRightSiblingKey()).putLong(node.getFirstChildKey());
+      if(lastChild != -2L) {
```
Comment (JohannesLichtenberger) [inline bundles/sirix-core/src/main/java/org/sirix/node/delegates/StructNodeDelegate.java:209]: Same as above. Maybe it should be a constant.

**Comment by JohannesLichtenberger** [comment]
That's pretty long PR, thanks a lot :+1:

File: bundles/sirix-core/src/main/java/org/sirix/node/delegates/StructNodeDelegate.java (hunk 2 above)
Comment (Raj-Datta-Manohar) [reply bundles/sirix-core/src/main/java/org/sirix/node/delegates/StructNodeDelegate.java:209]: Uh oh!! Will change

**Comment by JohannesLichtenberger** [comment]
@all-contributors please add @Raj-Datta-Manohar for code

**Comment by JohannesLichtenberger** [comment]
Thanks for the work. Looking forward to the tests :+1: and maybe further contributions if you like to and time permits :-)

**Comment by Raj-Datta-Manohar** [comment]
Sure:-), I plan to work on the open issues as well.

**Comment by JohannesLichtenberger** [comment]
BTW: do you have some database internals knowledge as you seem to work on Presto?

**Comment by Raj-Datta-Manohar** [comment]
Yes I do, I have also implemented a distributed in-memory database as a project!

**Comment by JohannesLichtenberger** [comment]
Is it your Ph.D. project? :+1:

That's great. I'm thinking about some sort of a simple group commit for the read-write transactions (single writer per resource): https://github.com/sirixdb/sirix/issues/296 and https://github.com/sirixdb/sirix/issues/299 or otherwise version the IndirectPages as well.

The issue is, that SirixDBs log-structure is also the main index, without the need for a WAL, as the main root page (the UberPage) can be swapped atomically.

The data pages are versioned, based on a sliding snapshot algorithm per default, meaning only page fragments are stored and currently a page has pointers to it's former versions. Thus, all nodes/records are stored in a new versionen of the page (a page-fragment), which have been changed in the current version or which would fall out of a sliding window. The offsets into the storage file currently are stored in the data page fragment itself, however they should reside in the PageReferences of the IndirectPages, such that Sirix can fetch the page fragments in parallel if the hardware permits this.

The basic problem I'm currently thinking about is how to avoid write amplification for small read-write transaction commits.

We could also version the IndirectPages, which form the simple trie and reading these fully in-memory (because it would involve up to n-page fragment fetches for each IndirectPage to reconstruct the page in-memory) or we could introduce a form of group commit. A group commit, however would complicate things, such as storing an author + commit message optionally during a commit, which is one of the main benefits of SirixDB, I think. In some sense SirixDB is very similar to Datomic, which also uses a persistent, durable log. However, they accumulate changes and store a log in DynamoDB for instance and they have a log + the real database I think.

I think our system otherwise is very similar to whats described in [1] from Caetano Sauer et al.

[1] http://www.vldb.org/pvldb/vol11/p2249-sauer.pdf

**Comment by JohannesLichtenberger** [comment]
Regarding Presto, I'm also working on Brackit, as Sebastian seems he doesn't have any time anymore: 

http://wwwlgis.informatik.uni-kl.de/cms/dbis/projects/brackit/mission/
http://wwwlgis.informatik.uni-kl.de/cms/fileadmin/publications/2013/Dissertation-Baechle.pdf

and the "fork" with some enhancements regarding JSON and temporal XPath axis: https://github.com/sirixdb/brackit

**Comment by Raj-Datta-Manohar** [comment]
No it was a personal project, I just completed my undergraduation:-)  
I think group commit is not a solution  
I will checkout the Brackit repo!!!

**Comment by JohannesLichtenberger** [comment]
Sounds great :-)

 BTW: it would be great to have some university research around the project again, as it's basically a fork of TreeTank, which is not maintained anymore. Thus, it started as a university project from 2006 - 2013 I think.

**Pull request merged**